dex/pair/src/safe_price_view.rs

```diff
@@ -0,0 +1,422 @@
+
+        if input_payment.token_identifier == first_token_id {
+            let output_amount =
+                input_payment.amount * weighted_second_token_reserve / weighted_first_token_reserve;
+            EsdtTokenPayment::new(second_token_id, 0, output_amount)
+        } else {
+            let output_amount =
+                input_payment.amount * weighted_first_token_reserve / weighted_second_token_reserve;
+            EsdtTokenPayment::new(first_token_id, 0, output_amount)
+        }
+    }
+
+    fn get_price_observation(
+        &self,
+        current_index: usize,
+        price_observations: &VecMapper<Self::Api, PriceObservation<Self::Api>>,
+        search_round: Round,
+    ) -> PriceObservation<Self::Api> {
+        require!(
+            !price_observations.is_empty(),
+            ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST
+        );
+
+        // Check if the requested price observation is the last one
+        let last_observation = price_observations.get(current_index);
+        if last_observation.recording_round == search_round {
+            return last_observation;
+        }
+
+        // Simulate a future price observation, based on the current reserves,
+        // in case the searched round is bigger than the last recording round
+        // The search round is limited to the current blockchain round
+        if last_observation.recording_round < search_round {
+            let current_round = self.blockchain().get_block_round();
+            let storage_cache = StorageCache::new(self);
+            return self.compute_new_observation(
+                core::cmp::min(search_round, current_round),
+                &storage_cache.first_token_reserve,
+                &storage_cache.second_token_reserve,
+                &last_observation,
+            );
+        }
+
+        let (mut price_observation, last_search_index) = self.price_observation_by_binary_search(
+            current_index,
+            price_observations,
+            search_round,
+        );
+
+        if price_observation.recording_round > 0 {
+            return price_observation;
+        }
+
+        price_observation = self.price_observation_by_linear_interpolation(
+            price_observations,
+            search_round,
+            last_search_index,
+        );
+
+        price_observation
+    }
+
+    fn get_oldest_price_observation(
+        &self,
+        current_index: usize,
+        price_observations: &VecMapper<Self::Api, PriceObservation<Self::Api>>,
+    ) -> PriceObservation<Self::Api> {
+        require!(
+            !price_observations.is_empty(),
+            ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST
+        );
+
+        // VecMapper index starts at 1
+        let mut oldest_observation_index = 1;
+        if price_observations.len() == MAX_OBSERVATIONS {
+            oldest_observation_index = (current_index % MAX_OBSERVATIONS) + 1
+        }
```

Comment:
unsure if correct.

Reply:
We use the modulo operator like so, because the starting index of the vec is 1, and not 0.